crates/extrinsics/src/upload.rs

```diff
@@ -239,3 +244,42 @@ pub struct UploadResult {
     pub code_stored: Option<CodeStored>,
     pub display_events: DisplayEvents,
 }
+
+#[derive(
+    Debug,
+    Clone,
+    Copy,
+    PartialEq,
+    Eq,
+    serde::Serialize,
+    scale::Decode,
+    scale::Encode,
+    scale_encode::EncodeAsType,
+)]
+#[encode_as_type(crate_path = "subxt::ext::scale_encode")]
+pub enum Determinism {
+    /// The execution should be deterministic and hence no indeterministic instructions
+    /// are allowed.
+    ///
+    /// Dispatchables always use this mode in order to make on-chain execution
+    /// deterministic.
+    Enforced,
+    /// Allow calling or uploading an indeterministic code.
+    ///
+    /// This is only possible when calling into `pallet-contracts` directly via
+    /// [`crate::Pallet::bare_call`].
+    ///
+    /// # Note
+    ///
+    /// **Never** use this mode for on-chain execution.
+    Relaxed,
+}
+
+/// A raw call to `pallet-contracts`'s `upload`.
+#[derive(Debug, scale_encode::EncodeAsType)]
+#[encode_as_type(crate_path = "subxt::ext::scale_encode")]
+pub struct UploadCode {
```

Comment:
I think that internal structs should be private, but here they are made public to keep it consistent with the rest of the code.